File: cli/src/languages/rust/rust.ts
import {
  Block,
  BlockCache,
  BlockTask,
  BlockTypes,
  DefaultOptions,
} from "src/core/block";
import { ValueType, ValueTypes } from "src/core/value";
import {
  assertUnreachable,
  camelCase,
  getTabs as get2Tabs,
  pascalCase,
} from "src/core/utilities";
import { Namespace, flatNamespace } from "src/core/namespace";
import { MapValue } from "src/core/generics";

const getTabs = (x: number) => get2Tabs(x * 2);

const blockOptionsToCode: (
  blockType: BlockTypes,
  level: number,
  options?: MapValue<(BlockCache | BlockTask)["values"]>["options"]
) => string = (bType, level, options) => {
  if (!options) {
    return "None";
  }
  switch (bType) {
    case BlockTypes.cache: {
      const o = options as MapValue<BlockCache["values"]>["options"];
      return o
        ? `Some(memorix_client_redis::MemorixOptionsCache {
${getTabs(level + 1)}expire: ${
            o.expire
              ? `Some(memorix_client_redis::MemorixOptionsCacheExpire {
${getTabs(level + 2)}value: ${o.expire.value},
${getTabs(level + 2)}is_in_ms: ${
                  o.expire.isInMs !== undefined
                    ? `Some(${o.expire.isInMs})`
                    : "None"
                },
${getTabs(level + 2)}extend_on_get: ${
                  o.expire.extendOnGet !== undefined
                    ? `Some(${o.expire.extendOnGet})`
                    : "None"
                },
${getTabs(level + 1)}})`
              : "None"
          },
${getTabs(level)}})`
        : "None";
    }
    case BlockTypes.task: {
      const o = options as MapValue<BlockTask["values"]>["options"];
      return o
        ? `Some(memorix_client_redis::MemorixOptionsTask {
${getTabs(level + 1)}take_newest: Some(${o.takeNewest ? "true" : "false"}),
${getTabs(level)}})`
        : "None";
    }
    case BlockTypes.enum:
    case BlockTypes.pubsub:
    case BlockTypes.model: {
      throw new Error("No options yet");
    }
    default:
      assertUnreachable(bType);
      return "";
  }
};

const valueToCode: (value: ValueType, isType: boolean) => string = (
  value,
  isType
) => {
  let valueRust;
  if (value.type === ValueTypes.simple) {
    if (value.name === "string") {
      valueRust = "String";
    } else if (value.name === "boolean") {
      valueRust = "bool";
    } else if (value.name === "float") {
      valueRust = "f32";
    } else if (value.name === "int") {
      valueRust = "i32";
    } else if (isType) {
      valueRust = `${value.name}`;
    } else {
      valueRust = `${value.name}`;
    }
  } else if (value.type === ValueTypes.array) {
    valueRust = `Vec<${valueToCode(value.value, isType)}>`;
  } else {
    throw new Error(
      "Shouldn't get here, all inline objects became models, maybe forgot 'flatBlocks()?'"
    );
  }

  return `${value.isOptional ? "Option<" : ""}${valueRust}${
    value.isOptional ? ">" : ""
  }`;
};

const blockToStruct: (block: Block) => string = (b) => {
  switch (b.type) {
    case BlockTypes.model:
    case BlockTypes.enum: {
      throw new Error(
        "Shouldn't get here, all inline objects became models, maybe forgot 'flatBlocks()?'"
      );
    }
    case BlockTypes.cache:
    case BlockTypes.pubsub:
    case BlockTypes.task: {
      const itemClass = {
        [BlockTypes.cache]: "MemorixCacheItem",
        [BlockTypes.pubsub]: "MemorixPubSubItem",
        [BlockTypes.task]: "MemorixTaskItem",
      }[b.type];
      const hasReturns = b.type === BlockTypes.task;

      return `${(
        Array.from(b.values.entries()) as [
          string,
          MapValue<typeof b["values"]>
        ][]
      )
        .map(([name, v]) => {
          return `${getTabs(1)}pub ${name}: memorix_client_redis::${itemClass}${
            v.key ? "" : "NoKey"
          }${
            hasReturns && !(v as MapValue<BlockTask["values"]>).returns
              ? "NoReturns"
              : ""
          }<${v.key ? `${valueToCode(v.key, true)}, ` : ""}${valueToCode(
            v.payload,
            true
          )}${
            (v as MapValue<BlockTask["values"]>).returns
              ? `, ${valueToCode(
                  (v as MapValue<BlockTask["values"]>).returns!,
                  true
                )}`
              : ""
          }>,`;
        })
        .join("\n")}`;
    }
    default:
      assertUnreachable(b);
      return "";
  }
};
const blockToCode: (block: Block) => string = (b) => {
  switch (b.type) {
    case BlockTypes.model:
      return `#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub struct ${b.name} {
${b.properties
  .map(
    (p) =>
      `${
        (p.value.type === ValueTypes.array ||
          p.value.type === ValueTypes.simple) &&
        p.value.isOptional
          ? `${getTabs(1)}#[serde(skip_serializing_if = "Option::is_none")]\n`
          : ""
      }${getTabs(1)}pub ${p.name === "type" ? "r#type" : p.name}: ${valueToCode(
        p.value,
        true
      )},`
  )
  .join(`\n`)}
}`;
    case BlockTypes.enum:
      return `#[allow(non_camel_case_types, clippy::upper_case_acronyms)]
#[derive(Clone, memorix_client_redis::Serialize, memorix_client_redis::Deserialize, PartialEq, std::fmt::Debug)]
pub enum ${b.name} {
${b.values.map((v) => `${getTabs(1)}${v},`).join(`\n`)}
}`;
    case BlockTypes.cache:
    case BlockTypes.pubsub:
    case BlockTypes.task: {
      const itemClass = {
        [BlockTypes.cache]: "MemorixCacheItem",
        [BlockTypes.pubsub]: "MemorixPubSubItem",
        [BlockTypes.task]: "MemorixTaskItem",
      }[b.type];
      const hasReturns = b.type === BlockTypes.task;

      return `${(
        Array.from(b.values.entries()) as [
          string,
          MapValue<typeof b["values"]>
        ][]
      )
        .map(([name, v]) => {
          return `${getTabs(3)}${name}: memorix_client_redis::${itemClass}${
            v.key ? "" : "NoKey"
          }${
            hasReturns && !(v as MapValue<BlockTask["values"]>).returns
              ? "NoReturns"
              : ""
          }::new(
${getTabs(4)}memorix_base.clone(),
${getTabs(4)}"${name}".to_string(),${
            b.type === BlockTypes.cache || b.type === BlockTypes.task
              ? `
${getTabs(4)}${blockOptionsToCode(
                  b.type,
                  4,
                  {
                    [BlockTypes.cache]: (v as MapValue<BlockCache["values"]>)
                      .options,
                    [BlockTypes.task]: (v as MapValue<BlockTask["values"]>)
                      .options,
                  }[b.type]
                )}`
              : ""
          }
${getTabs(3)}),`;
        })
        .join("\n")}`;
    }
    default:
      assertUnreachable(b);
      return "";
  }
};

const defaultOptionsToCode: (defaultOptions?: DefaultOptions) => string = (
  defaultOptions
) => {
  if (!defaultOptions) {
    return "None";
  }
  return `Some(memorix_client_redis::MemorixOptions {
${getTabs(4)}cache: ${blockOptionsToCode(
    BlockTypes.cache,
    4,
    defaultOptions.cache
  )},
${getTabs(4)}task: ${blockOptionsToCode(
    BlockTypes.task,
    4,
    defaultOptions.task
  )},
${getTabs(3)}})`;
};

const namespaceToCode: (
  namespace: Namespace,
  nameTree?: string[]
) => {
  code: string;
  importBase: boolean;
  importCache: boolean;
  importPubSub: boolean;
  importTask: boolean;
  importEnum: boolean;
} = (namespace, nameTree = []) => {
  const hasEnum = namespace.enums.size > 0;
  const hasCache = !!namespace.cache;
  const hasPubsub = !!namespace.pubsub;
  const hasTask = !!namespace.task;
  const hasApi = hasCache || hasPubsub || hasTask || !!namespace.defaultOptions;

  const subSamespaces = Array.from(namespace.subNamespacesByName.keys()).map(
    (name) =>
      namespaceToCode(namespace.subNamespacesByName.get(name)!, [
        ...nameTree,
        name,
      ])
  );
  const nameCamel = nameTree.map((x) => camelCase(x)).join("");
  const namePascal = nameTree.map((x) => pascalCase(x)).join("_");

  const code = ([] as string[])
    .concat(Array.from(namespace.enums.values()).map(blockToCode))
    .concat(Array.from(namespace.models.values()).map(blockToCode))
    .concat(subSamespaces.map((x) => x.code))
    .concat(
      hasCache
        ? `#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixCache${nameCamel} {
${blockToStruct(namespace.cache!)}
}

impl MemorixCache${nameCamel} {
${getTabs(1)}fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
${getTabs(2)}Self {
${blockToCode(namespace.cache!)}
${getTabs(2)}}
${getTabs(1)}}
}`
        : []
    )
    .concat(
      hasPubsub
        ? `#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixPubSub${nameCamel} {
${blockToStruct(namespace.pubsub!)}
}

impl MemorixPubSub${nameCamel} {
${getTabs(1)}fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
${getTabs(2)}Self {
${blockToCode(namespace.pubsub!)}
${getTabs(2)}}
${getTabs(1)}}
}`
        : []
    )
    .concat(
      hasTask
        ? `#[derive(Clone)]
#[allow(non_snake_case)]
pub struct MemorixTask${nameCamel} {
${blockToStruct(namespace.task!)}
}

impl MemorixTask${nameCamel} {
${getTabs(1)}fn new(memorix_base: memorix_client_redis::MemorixBase) -> Self {
${getTabs(2)}Self {
${blockToCode(namespace.task!)}
${getTabs(2)}}
${getTabs(1)}}
}`
        : []
    )
    .concat(
      hasApi || namespace.subNamespacesByName.size !== 0
        ? `#[derive(Clone)]
#[allow(non_snake_case)]
pub struct Memorix${nameCamel} {
${Array.from(namespace.subNamespacesByName.keys())
  .map(
    (namespaceName) =>
      `${getTabs(3)}pub ${namespaceName}: Memorix${nameCamel}${camelCase(
        namespaceName
      )},`
  )
  .join("\n")}${
            Array.from(namespace.subNamespacesByName.keys()).length !== 0
              ? "\n"
              : ""
          }
${([] as string[])
  .concat(hasCache ? `${getTabs(1)}pub cache: MemorixCache${nameCamel},` : [])
  .concat(
    hasPubsub ? `${getTabs(1)}pub pubsub: MemorixPubSub${nameCamel},` : []
  )
  .concat(hasTask ? `${getTabs(1)}pub task: MemorixTask${nameCamel},` : [])
  .join("\n")}
}

const MEMORIX_${
            namePascal ? `${namePascal}_` : ""
          }NAMESPACE_NAME_TREE: &[&str] = &[${nameTree
            .map((x) => `"${x}"`)
            .join(", ")}];

impl Memorix${nameCamel} {
${
  nameTree.length === 0
    ? `${getTabs(
        1
      )}pub async fn new(redis_url: &str) -> Result<Memorix${nameCamel}, Box<dyn std::error::Error + Sync + Send>> {
${getTabs(2)}let memorix_base = memorix_client_redis::MemorixBase::new(
${getTabs(3)}redis_url,
${getTabs(3)}MEMORIX_${namePascal ? `${namePascal}_` : ""}NAMESPACE_NAME_TREE,
${getTabs(3)}${defaultOptionsToCode(namespace.defaultOptions)}
${getTabs(2)}).await?;`
    : `${getTabs(
        1
      )}pub fn new(other: memorix_client_redis::MemorixBase) -> Result<Memorix${nameCamel}, Box<dyn std::error::Error + Sync + Send>> {
${getTabs(2)}let memorix_base = memorix_client_redis::MemorixBase::from(
${getTabs(3)}other,
${getTabs(3)}MEMORIX_${namePascal ? `${namePascal}_` : ""}NAMESPACE_NAME_TREE,
${getTabs(3)}${defaultOptionsToCode(namespace.defaultOptions)}
${getTabs(2)});`
}
${getTabs(2)}Ok(Self {
${Array.from(namespace.subNamespacesByName.keys())
  .map(
    (namespaceName) =>
      `${getTabs(3)}${namespaceName}: Memorix${nameCamel}${camelCase(
        namespaceName
      )}::new(memorix_base.clone())?,`
  )
  .join("\n")}${
            Array.from(namespace.subNamespacesByName.keys()).length !== 0
              ? "\n"
              : ""
          }
${([] as string[])
  .concat(
    hasCache
      ? `${getTabs(
          3
        )}cache: MemorixCache${nameCamel}::new(memorix_base.clone()),`
      : []
  )
  .concat(
    hasPubsub
      ? `${getTabs(
          3
        )}pubsub: MemorixPubSub${nameCamel}::new(memorix_base.clone()),`
      : []
  )
  .concat(
    hasTask
      ? `${getTabs(3)}task: MemorixTask${nameCamel}::new(memorix_base.clone()),`
      : []
  )
  .join("\n")}
${getTabs(2)}})
${getTabs(1)}}
}`
        : []
    )
    .join("\n\n\n");

  return {
    code,
    importBase: hasApi || subSamespaces.some((x) => x.importBase),
    importEnum: hasEnum || subSamespaces.some((x) => x.importEnum),
    importCache: hasCache || subSamespaces.some((x) => x.importCache),
    importPubSub: hasPubsub || subSamespaces.some((x) => x.importPubSub),
    importTask: hasTask || subSamespaces.some((x) => x.importTask),
  };
};

export const codegen: (namespaces: Namespace) => string = (
  unflattenNamespace
) => {
  const namespace = flatNamespace(unflattenNamespace);
  const { code } = namespaceToCode(namespace);

  const importCode = ([] as string[])
    .concat([`extern crate memorix_client_redis;`])
    .join("\n");
  return `#![allow(dead_code)]
${importCode}

${code}`;
};

File: cli/src/languages/rust.rs
use crate::{
    flat_schema::{FlatExportNamespace, FlatExportSchema, FlatTypeItem, TypeItemObject},
    parser::{Engine, Value, ALL_CACHE_OPERATIONS, ALL_PUBSUB_OPERATIONS, ALL_TASK_OPERATIONS},
};

fn indent(level: usize) -> String {
    "    ".repeat(level)
}
fn flat_type_item_to_code(flat_type_item: &FlatTypeItem) -> String {
    match flat_type_item {
        FlatTypeItem::Optional(x) => format!("Option<{}>", flat_type_item_to_code(x)).to_string(),
        FlatTypeItem::Array(x) => format!("Vec<{}>", flat_type_item_to_code(x)).to_string(),
        FlatTypeItem::Reference(x) => x.clone(),
        FlatTypeItem::Boolean => "bool".to_string(),
        FlatTypeItem::String => "String".to_string(),
        FlatTypeItem::U32 => "u32".to_string(),
        FlatTypeItem::U64 => "u64".to_string(),
        FlatTypeItem::I32 => "i32".to_string(),
        FlatTypeItem::I64 => "i64".to_string(),
        FlatTypeItem::F32 => "f32".to_string(),
        FlatTypeItem::F64 => "f64".to_string(),
    }
}

fn value_to_code(value: &Value) -> String {
    match value {
        Value::String(x) => format!("\"{x}\".to_string()"),
        Value::Env(x) => {
            format!("std::env::var(\"{x}\").expect(\"missing environment variable {x}\")")
        }
    }
}

fn type_item_object_to_code(name: &str, type_item_object: &TypeItemObject) -> String {
    let property_indent = indent(1);
    format!(
        r#"
#[memorix_client_redis::serialization]
#[derive(Clone, PartialEq, std::fmt::Debug)]
pub struct {name} {{
{}
}}
"#,
        type_item_object
            .properties
            .iter()
            .map(|(property_name, flat_type_item)| format!(
                "{property_indent}pub {}: {},",
                match property_name == "type" {
                    true => "r#type",
                    false => property_name,
                },
                flat_type_item_to_code(flat_type_item)
            )
            .to_string())
            .collect::<Vec<_>>()
            .join("\n")
    )
    .to_string()
}

fn namespace_to_code(
    namespace: &FlatExportNamespace,
    name_tree: Vec<String>,
    engine: &Engine,
) -> String {
    let mut result = String::from("");
    let base_indent = indent(name_tree.len());

    for (name, values) in &namespace.enum_items {
        result.push_str(&format!(
            r#"
{base_indent}#[allow(non_camel_case_types, clippy::upper_case_acronyms)]
{base_indent}#[memorix_client_redis::serialization]
{base_indent}#[derive(Clone, PartialEq, std::fmt::Debug)]
{base_indent}pub enum {name} {{
{}
}}

"#,
            values
                .iter()
                .map(|x| format!("{base_indent}    {x},"))
                .collect::<Vec<_>>()
                .join("\n")
        ));
    }
    for (name, type_item_object) in &namespace.type_item_objects {
        result.push_str(&format!(
            "{}\n",
            type_item_object_to_code(name.as_str(), type_item_object)
        ));
    }
    for (name, flat_type_item) in &namespace.flat_type_items {
        result.push_str(&format!(
            "{base_indent}pub type {name} = {};\n\n",
            flat_type_item_to_code(flat_type_item)
        ));
    }
    for (name, sub_namespace) in &namespace.namespaces {
        result.push_str(&format!(
            r#"{base_indent}pub mod {name} {{
{base_indent}    use super::*;

{namespace_content}
{base_indent}}}
"#,
            name = stringcase::snake_case(name),
            namespace_content = namespace_to_code(
                sub_namespace,
                name_tree
                    .clone()
                    .into_iter()
                    .chain(std::iter::once(name.clone()))
                    .collect(),
                engine,
            ),
        ));
    }
    if !namespace.cache_items.is_empty() {
        result.push_str(&format!(
            r#"{base_indent}#[derive(Clone)]
{base_indent}#[allow(non_snake_case)]
{base_indent}pub struct MemorixCache {{
{struct_content}
{base_indent}}}

{base_indent}impl MemorixCache {{
{base_indent}    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Result<Self, Box<dyn std::error::Error + Sync + Send>> {{
{base_indent}        Ok(Self {{
{impl_content}
{base_indent}        }})
{base_indent}    }}
{base_indent}}}

"#,
            struct_content =
                namespace
                    .cache_items
                    .iter()
                    .map(|(name, item)| {
                        let payload = flat_type_item_to_code(&item.payload);
                        format!(
                        "{base_indent}    pub {name}: memorix_client_redis::MemorixCacheItem{key}{api}>,",
                        key = match &item.key {
                            None => format!("NoKey<{payload}, "),
                            Some(key) => {
                                let key = flat_type_item_to_code(key);
                                format!("<{key}, {payload}, ")
                            }
                        },
                        api = ALL_CACHE_OPERATIONS.iter().map(|x| match item.expose.contains(x) {
                            true => "memorix_client_redis::Expose",
                            false => "memorix_client_redis::Hide",
                        }).collect::<Vec<_>>().join(", ")
                    )
                    })
                    .collect::<Vec<_>>()
                    .join("\n"),
            impl_content = namespace
                .cache_items
                .iter()
                .map(|(name, item)| {
                    let options = format!(
                        r#"Some(memorix_client_redis::MemorixCacheOptions {{
{content}
{base_indent}                }})"#,
                        content = [
                            format!(
                                "{base_indent}                    ttl: {},",
                                item.ttl.as_ref().map_or("None".to_string(), |x| format!(
                                    "Some({})",
                                    value_to_code(x)
                                ))
                            ),
                            format!(
                                "{base_indent}                    extend_on_get: {},",
                                item.extend_on_get.as_ref().map_or(
                                    "None".to_string(),
                                    |x| format!("Some({})", value_to_code(x))
                                )
                            ),
                        ]
                        .into_iter()
                        .collect::<Vec<_>>()
                        .join("\n")
                    );
                    format!(
                        r#"{base_indent}            {name}: memorix_client_redis::MemorixCacheItem{key}::new(
{base_indent}                memorix_base.clone(),
{base_indent}                "{name}".to_string(),
{base_indent}                {options},
{base_indent}            )?,"#,
                        key = match &item.key {
                            None => "NoKey",
                            Some(_) => "",
                        },
                        options = options,
                    )
                })
                .collect::<Vec<_>>()
                .join("\n")
        ));
    }
    if !namespace.pubsub_items.is_empty() {
        result.push_str(&format!(
            r#"{base_indent}#[derive(Clone)]
{base_indent}#[allow(non_snake_case)]
{base_indent}pub struct MemorixPubSub {{
{struct_content}
{base_indent}}}

{base_indent}impl MemorixPubSub {{
{base_indent}    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Result<Self, Box<dyn std::error::Error + Sync + Send>> {{
{base_indent}        Ok(Self {{
{impl_content}
{base_indent}        }})
{base_indent}    }}
{base_indent}}}

"#,
            struct_content =
                namespace
                    .pubsub_items
                    .iter()
                    .map(|(name, item)| {
                        let payload = flat_type_item_to_code(&item.payload);
                        format!(
                        "{base_indent}    pub {name}: memorix_client_redis::MemorixPubSubItem{key}{api}>,",
                        key = match &item.key {
                            None => format!("NoKey<{payload}, "),
                            Some(key) => {
                                let key = flat_type_item_to_code(key);
                                format!("<{key}, {payload}, ")
                            }
                        },
                        api = ALL_PUBSUB_OPERATIONS.iter().map(|x| match item.expose.contains(x) {
                            true => "memorix_client_redis::Expose",
                            false => "memorix_client_redis::Hide",
                        }).collect::<Vec<_>>().join(", ")
                    )
                    })
                    .collect::<Vec<_>>()
                    .join("\n"),
            impl_content = namespace
                .pubsub_items
                .iter()
                .map(|(name, item)| {
                    format!(
                        r#"{base_indent}            {name}: memorix_client_redis::MemorixPubSubItem{key}::new(
{base_indent}                memorix_base.clone(),
{base_indent}                "{name}".to_string(),
{base_indent}            )?,"#,
                        key = match &item.key {
                            None => "NoKey",
                            Some(_) => "",
                        },
                    )
                })
                .collect::<Vec<_>>()
                .join("\n")
        ));
    }
    if !namespace.task_items.is_empty() {
        result.push_str(&format!(
            r#"{base_indent}#[derive(Clone)]
{base_indent}#[allow(non_snake_case)]
{base_indent}pub struct MemorixTask {{
{struct_content}
{base_indent}}}

{base_indent}impl MemorixTask {{
{base_indent}    fn new(memorix_base: memorix_client_redis::MemorixBase) -> Result<Self, Box<dyn std::error::Error + Sync + Send>> {{
{base_indent}        Ok(Self {{
{impl_content}
{base_indent}        }})
{base_indent}    }}
{base_indent}}}

"#,
            struct_content =
                namespace
                    .task_items
                    .iter()
                    .map(|(name, item)| {
                        let payload = flat_type_item_to_code(&item.payload);
                        format!(
                        "{base_indent}    pub {name}: memorix_client_redis::MemorixTaskItem{key}{api}>,",
                        key = match &item.key {
                            None => format!("NoKey<{payload}, "),
                            Some(key) => {
                                let key = flat_type_item_to_code(key);
                                format!("<{key}, {payload}, ")
                            }
                        },
                        api = ALL_TASK_OPERATIONS.iter().map(|x| match item.expose.contains(x) {
                            true => "memorix_client_redis::Expose",
                            false => "memorix_client_redis::Hide",
                        }).collect::<Vec<_>>().join(", ")
                    )
                    })
                    .collect::<Vec<_>>()
                    .join("\n"),
            impl_content = namespace
                .task_items
                .iter()
                .map(|(name, item)| {
                    let options = format!(
                        r#"Some(memorix_client_redis::MemorixTaskOptions {{
{content}
{base_indent}                }})"#,
                        content =
                            [format!(
                                "{base_indent}                    queue_type: {},",
                                item.queue_type.as_ref().map_or(
                                    "None".to_string(),
                                    |x| format!("Some({})", value_to_code(x))
                                )
                            ),]
                            .into_iter()
                            .collect::<Vec<_>>()
                            .join("\n")
                    );
                    format!(
                        r#"{base_indent}            {name}: memorix_client_redis::MemorixTaskItem{key}::new(
{base_indent}                memorix_base.clone(),
                "{name}".to_string(),
{base_indent}                {options},
{base_indent}            )?,"#,
                        key = match &item.key {
                            None => "NoKey",
                            Some(_) => "",
                        },
                        options = options,
                    )
                })
                .collect::<Vec<_>>()
                .join("\n")
        ));
    }

    result.push_str(&format!(
        r#"{base_indent}#[derive(Clone)]
{base_indent}#[allow(non_snake_case)]
{base_indent}pub struct Memorix {{
{struct_content}
{base_indent}}}

{base_indent}const MEMORIX_NAMESPACE_NAME_TREE: &[&str] = &[{name_tree}];

{base_indent}impl Memorix {{
{impl_new}
{base_indent}        Ok(Self {{
{impl_content}
{base_indent}        }})
{base_indent}    }}
{base_indent}}}
"#,
        name_tree = name_tree
            .iter()
            .map(|x| format!("\"{}\"", x))
            .collect::<Vec<_>>()
            .join(", "),
        struct_content = [
            namespace
                .namespaces
                .iter()
                .map(|(name, _)| format!(
                    "{base_indent}    pub {name}: {namespace_name}::Memorix,",
                    namespace_name = stringcase::snake_case(name)
                ))
                .collect::<Vec<_>>()
                .join("\n"),
            [
                (!namespace.cache_items.is_empty())
                    .then(|| format!("{base_indent}    pub cache: MemorixCache,")),
                (!namespace.pubsub_items.is_empty())
                    .then(|| format!("{base_indent}    pub pubsub: MemorixPubSub,")),
                (!namespace.task_items.is_empty())
                    .then(|| format!("{base_indent}    pub task: MemorixTask,")),
            ]
            .into_iter()
            .flatten()
            .collect::<Vec<_>>()
            .join("\n")
        ]
        .into_iter()
        .filter(|x| !x.is_empty())
        .collect::<Vec<_>>()
        .join("\n\n"),
        impl_new = match name_tree.is_empty() {
            true => format!(r#"{base_indent}    pub async fn new() -> Result<Memorix, Box<dyn std::error::Error + Sync + Send>> {{
{base_indent}        let memorix_base = memorix_client_redis::MemorixBase::new(
{base_indent}            {redis_url},
{base_indent}            MEMORIX_NAMESPACE_NAME_TREE
{base_indent}        )
{base_indent}        .await?;"#, redis_url= match engine {
            Engine::Redis(x) => format!("&{}", value_to_code(x)),
            }),
            false => format!(
                r#"{base_indent}    pub fn new(
{base_indent}        other: memorix_client_redis::MemorixBase,
{base_indent}    ) -> Result<Memorix, Box<dyn std::error::Error + Sync + Send>> {{
{base_indent}        let memorix_base = memorix_client_redis::MemorixBase::from(
{base_indent}            other,
{base_indent}            MEMORIX_NAMESPACE_NAME_TREE
{base_indent}        );"#),
        },
        impl_content = [
            namespace
                .namespaces
                .iter()
                .map(|(name, _)| format!(
                    "{base_indent}            {name}: {namespace_name}::Memorix::new(memorix_base.clone())?,",
                    namespace_name = stringcase::snake_case(name),
                ))
                .collect::<Vec<_>>()
                .join("\n"),
            [
                (!namespace.cache_items.is_empty()).then(|| format!(
                    "{base_indent}           cache: MemorixCache::new(memorix_base.clone())?,"
                )),
                (!namespace.pubsub_items.is_empty()).then(|| format!(
                    "{base_indent}            pubsub: MemorixPubSub::new(memorix_base.clone())?,"
                )),
                (!namespace.task_items.is_empty()).then(|| format!(
                    "{base_indent}            task: MemorixTask::new(memorix_base.clone())?,"
                )),
            ]
            .into_iter()
            .flatten()
            .collect::<Vec<_>>()
            .join("\n")
        ]
        .into_iter()
        .filter(|x| !x.is_empty())
        .collect::<Vec<_>>()
        .join("\n\n")
    ));
    result
}

pub fn codegen(flat_export_schema: &FlatExportSchema) -> String {
    format!(
        r#"#![allow(dead_code)]
extern crate memorix_client_redis;

{}"#,
        namespace_to_code(
            &flat_export_schema.global_namespace,
            vec![],
            &flat_export_schema.engine,
        )
    )
    .to_string()
}
